### PR TITLE
DA_Week11_복습과제_손소현

### PR DESCRIPTION
**Pull request by sonso1598**

### CODE REVIEW

**Comment by lej3034** [comment]
소현님 안녕하세요 복습과제 피드백 드리겠습니다!

1-6. 4) 군집별 평균 실루엣 계수 값을 구하는 문제
cd_copy.groupby('cluster')['silhouette_coeff'].mean()

1-8. estimate_bandwidth() 사용해 최적 bandwidth 값 구하는 문제
from sklearn.cluster import estimate_bandwidth
best_bandwidth = estimate_bandwidth(cd_copy2)
print('best_bandwidth 값:', round(best_bandwidth,3))

1-9. 최적 bandwidth를 평균 이동 입력값으로 적용해 다시 fit하고 cluster labels 유형을 확인
meanshift= MeanShift(bandwidth=best_bandwidth)
cluster_labels = meanshift.fit_predict(cd_copy2)
print('cluster labels 유형:',np.unique(cluster_labels))    

2-10. 사이킷런의 DBSCAN을 활용하여 x를 fit
model = DBS_clustering.fit(x)

2-11. labels_을 출력하여 확인
label=model.labels_

2-12. 다음 코드를 실행하여 DBSCAN의 결과를 확인
db->DBS_clustering 로 수정

4-2. 사이킷런의 베이즈가우시안 혼합 모델을 활용하여 x를 fit
from sklearn.mixture import BayesianGaussianMixture
bgm = BayesianGaussianMixture(n_components=20, random_state=33)
bgm.fit(x)

4-3. 불필요한 클러스터의 가중치를 0으로 잘 처리했는지 확인
np.round(bgm.weights_, 2)

4-4. 다음의 코드를 실행하여 visualize_cluster_plot 함수를 만들고 군집화 결과를 시각화
bgm_label = bgm.fit(x).predict(x)
clusterDF['bgm_label'] = bgm_label
visualize_cluster_plot(bgm, clusterDF, 'bgm_label',iscenter=False)

5-1. pca 모듈로 X_train 분산을 99% 로 줄여주세요
from sklearn.decomposition import PCA
pca = PCA(0.99)
X_train_pca = pca.fit_transform(X_train)

5-2. AIC, BIC 점수를 비교해 사이킷런의 GaussianMixture 모듈의 최적의 클러스터 개수를 찾기
from sklearn.mixture import GaussianMixture
n_components = np.arange(1, 16)
models = [GaussianMixture(n, random_state=33).fit(X_train_pca) for n in n_components]
aics = [model.aic(X_train_pca) for model in models]
bics = [model.bic(X_train_pca) for model in models]
plt.plot(n_components, aics, 'rs--')
plt.plot(n_components, bics, 'bo-')
plt.legend(['AIC', 'BIC'])
plt.show()

5-3. 최적 n_component 값을 넣어 gmm을 fit 시키고 알고리즘이 수렴했는지 확인
gmm = GaussianMixture(n_components=3, random_state=33)
gmm.fit(X_train_pca)
print(gmm.converged_)

과제하시느라 수고 많으셨습니다!!
